frontend/src/components/common/Resource/PortForward.tsx

```diff
@@ -1,78 +1,167 @@
+      if (port.name === namedPort) {
+        portNumber = port.containerPort;
+        return false;
+      }
+    });
+    return true;
+  });
+  return portNumber;
+}
+
+function getPodsSelectorFilter(service?: Service) {
+  if (!service) {
+    return '';
+  }
+  const selector = service?.jsonData.spec?.selector;
+  if (selector) {
+    return Object.keys(service?.jsonData.spec?.selector)
+      .map(item => `${item}=${selector[item]}`)
+      .join(',');
+  }
+  return '';
 }
 
 export default function PortForward(props: PortForwardProps) {
-  const { isPod, containerPort, namespace, name, isPodRunning } = props;
+  const { containerPort, resource } = props;
+  const isPod = resource?.metadata?.kind !== 'Service';
+  const service = !isPod ? (resource as Service) : undefined;
+  const namespace = resource?.metadata?.namespace;
+  const name = resource?.metadata?.name;
   const [isCurrentPodOrServicePortForwarding, setIsCurrentPodOrServicePortForwarding] =
     React.useState(false);
   const [error, setError] = React.useState(null);
   const [portForward, setPortForward] = React.useState<PortForwardState | null>(null);
   const [loading, setLoading] = React.useState(false);
   const cluster = getCluster();
-  const { t } = useTranslation('frequent');
+  const { t } = useTranslation(['frequent', 'resource']);
+  const [pods, podsFetchError] = Pod.useList({ labelSelector: getPodsSelectorFilter(service) });
+  const massagedContainerPort =
+    typeof containerPort === 'string' && isNaN(parseInt(containerPort))
+      ? getPortNumberFromPortName(pods[0].spec.containers, containerPort)
+      : containerPort;
+
+  React.useEffect(() => {
+    if (!cluster) {
+      return;
+    }
+    listPortForward(cluster).then(result => {
+      const portForwards = result;
+      if (!portForwards || portForwards.length === 0) {
+        return;
+      }
+      for (const item of portForwards) {
+        if (
+          (item.namespace === namespace || item.serviceNamespace === namespace) &&
+          (item.pod === name || item.service === name) &&
+          item.cluster === cluster &&
+          item.targetPort === massagedContainerPort.toString()
+        ) {
+          setIsCurrentPodOrServicePortForwarding(true);
+          setPortForward(item);
+          return;
+        }
+      }
+      const massagedPortForwards = [...portForwards];
+      const portForwardsInStorage = localStorage.getItem(PORT_FORWARDS_STORAGE_KEY);
+      const parsedPortForwards = JSON.parse(portForwardsInStorage || '[]');
+      parsedPortForwards.forEach((portforward: any) => {
+        const isPortForwardInStorage = portForwards.find((pf: any) => pf.id === portforward.id);
+        if (!isPortForwardInStorage) {
+          portforward.status = PORT_FORWARD_STOP_STATUS;
+          massagedPortForwards.push(portforward);
+        }
+      });
+      localStorage.setItem(PORT_FORWARDS_STORAGE_KEY, JSON.stringify(massagedPortForwards));
```

Comment:
I noticed another issue with the PR and it is related to this block:
1. Create a container port;
2. Restart the backend;
3. Go to the pod that had the container port -> it'll allow the user to start a new port.

This is related to how the ports concept is being managed here (where it's taking the backend's response as the source of truth).

Instead, it's the ports in the *local storage* (used created) that need to be the source of truth in what comes to what ports we have forwarded, as that reflects the user's intentions; the *backend* should be the source of the *state* for those ports.

So the logic should be rewritten as:
1. Get ports from the local storage;
2. Set the current portforward if it matches the current pod/service.
3. Set up the port's status from the backend response.

Reply:
fixed by checking this later on a data where we have storage and backend portforward data, so there is no inconsistency